django/db/models/indexes.py

```diff
@@ -0,0 +1,120 @@
+from __future__ import unicode_literals
+
+import hashlib
+
+from django.utils.encoding import force_bytes
+
+__all__ = ['Index']
+
+# The max length of the names of the indexes (restricted to 30 due to Oracle)
+MAX_NAME_LENGTH = 30
+
+
+class Index(object):
+
+    index_type = 'idx'
+
+    def __init__(self, fields=[], name=None):
+        if not fields:
+            raise ValueError('At least one field is required to define an index.')
+        self.fields = fields
+        self._name = name or ''
+        if self._name:
+            self.normalise_name()
+            if len(self._name) > MAX_NAME_LENGTH:
+                raise ValueError("Index names cannot be longer than %s characters." % MAX_NAME_LENGTH)
+
+    @property
+    def name(self):
+        if not self._name:
+            self._name = self.get_name()
+            self.normalise_name()
+        return self._name
+
+    def normalise_name(self):
+        # Name shouldn't start with an underscore (Oracle hates this), so prepend D if we need to
+        if self._name[0] == "_":
+            self._name = 'D%s' % self._name[1:]
+        # It can't start with a number on Oracle, so prepend D if we need to
+        elif self._name[0].isdigit():
+            self._name = 'D%s' % self._name[1:]
+
+    def create_sql(self, schema_editor):
+        columns = [field for field in self.fields]
+        fields = [self.model._meta.get_field(field) for field in self.fields]
+        if len(fields) == 1 and fields[0].db_tablespace:
+            tablespace_sql = schema_editor.connection.ops.tablespace_sql(fields[0].db_tablespace)
+        elif self.model._meta.db_tablespace:
+            tablespace_sql = schema_editor.connection.ops.tablespace_sql(self.model._meta.db_tablespace)
+        else:
+            tablespace_sql = ""
+        if tablespace_sql:
+            tablespace_sql = " " + tablespace_sql
+
+        quote_name = schema_editor.quote_name
+        return schema_editor.sql_create_index % {
+            'table': quote_name(self.model._meta.db_table),
+            'name': quote_name(self.name),
+            'columns': ', '.join(quote_name(column) for column in columns),
+            'extra': tablespace_sql,
+        }
+
+    def remove_sql(self, schema_editor):
+        quote_name = schema_editor.quote_name
+        return schema_editor.sql_delete_index % {
+            'table': quote_name(self.model._meta.db_table),
+            'name': quote_name(self.name),
+        }
+
+    def deconstruct(self):
+        """
+        Return a 3-tuple of class import path, positional arguments, and
```

Comment:
I think we can drop this docstring. We don't need to repeat it for every object that implements deconstruct().